### DISCUSSION TITLE
Improve loading UX when columns are fetched dynamically

### DISCUSSION THREAD

**Discussion by aarjithn** [KevinVandy/material-react-table#1001] [Ideas]
Hi,

~~The examples currently have column definitions within component, but in many real world applications, columns too get loaded via API.~~

This is the current behaviour by default, initial state of empty array which gets replaced with columns on load

``` ts
const [columns, setColumns] = useState<MRT_ColumnDef<User>[]>([]);
```

![mrt-dynamic-columns-loading-current](https://github.com/KevinVandy/material-react-table/assets/10304259/890806e5-1f4c-4c27-9c51-51ed4ca5b6fc)

This could be made significantly better with some hacky initial state when you know a rough number of columns to be expected.

``` ts
  const [columns, setColumns] = useState<MRT_ColumnDef<User>[]>(
    [{header: ' '}, {header: '  '}, {header: '   '}, {header: '    '}]
  );
```

![mrt-dynamic-columns-loading-hack](https://github.com/KevinVandy/material-react-table/assets/10304259/f1df5694-37d4-42f8-a26e-bbf7a0fa5564)

I have made  a sandbox as a demo:  https://stackblitz.com/edit/mrt-loading-ux?file=src%2FTS.tsx

Would be great to have some sort of official support to render a given number of columns, with a skeleton in column header cell while columns are loading.

Something like array of X nulls as initial state, and in that case show X columns with skeletons in header.

What do you think?

**Comment by KevinVandy** [comment]
fyi, there's a dedicated dynamic columns example https://www.material-react-table.com/docs/examples/dynamic-columns

**Reply by aarjithn** [reply]
Oops sorry I missed that. Since the columns load fast in that example it doesn't look bad. 
Also it uses react query so only issue is first time load - toggling between tabs don't make it noticeable.

If there is a delay though, maybe there is scope to make the loading state better?
